LLM Tools (#2164)
home-assistant/developers.home-assistant: docs/llm_tools.md

---
title: "LLM Tools"
sidebar_label: "LLM Tools"
---

> LLM tools extend the functionality of OpenAI Conversation, Google Generative AI, and possibly other future LLM integrations. You can define your own LLM tool that these integrations can use.

Reviewer: ```suggestion
LLM tools extend the functionality of LLM integrations such as OpenAI Conversation, Google Generative AI, and others. You can define your own LLM tool that these integrations can use.
```

> There are 3 options to define your own tool.
> 
> ## Option 1. Intent
> 
> Create and register your own intent handler. Intents are automatically registered as LLM tools. This method allows to use the same code for other integrations that support intents.

Reviewer: Instead of defining these as 3 options, lets document the APIs and also just say that intents are automatically registered.

```suggestion
All [Intents](/docs/intent_index) are automatically registered as LLM tools. This page describes the APIs for registering additional tools.
```

Reviewer: The tool registration docs can be omitted given we're reducing scope of the PR to only register intents.

> ## Option 2. Tool class

Reviewer: ```suggestion
## Tool
```

> Define a child class of `homeassistant.helpers.llm.Tool` and register an object of that class. This it the most flexible method. Example:
> 

Reviewer: ```suggestion
The most flexible method to define a tool is by registering a child class of `homeassistant.helpers.llm.Tool`. Example:

```

```python
from typing import Any

import voluptuous as vol

from homeassistant.core import HomeAssistant
from homeassistant.helpers import llm

class MyTool(llm.Tool):
    """LLM multiplication helper tool."""

    name = "multiply"
    description = "Return a product of two integers"
    parameters = vol.Schema(
        {
            vol.Required("a", description="multiplicand"): vol.Coerce(int),
            vol.Required("b", description="multiplier"): vol.Coerce(int),
        }
    )

    async def async_call(self, hass: HomeAssistant, tool_input: llm.ToolInput) -> Any:
        """Call the function."""
        args = tool_input.tool_args
        return {"result": args["a"] * args["b"]}

my_tool = MyTool()

> llm.async_register_tool(hass, my_tool)

Reviewer: I think it may be preferred to use a more realistic registration like in the example in  https://developers.home-assistant.io/docs/dev_101_services/ or https://developers.home-assistant.io/docs/intent_handling

(I don't think its common to see inline code like this)

```

Let's unpack.

The `llm.Tool` class has the following user-defined attributes:

| Name                | Type       | Description                                                                                                    |
|---------------------|------------|----------------------------------------------------------------------------------------------------------------|
| `name`              | string     | The name of the tool. Required.                                                                                |
| `description`       | string     | Description of the tool to help the LLM understand when and how it should be called. Optional but recommended. |
| `parameters`        | vol.Schema | The voluptuous schema of the parameters. Defaults to vol.Schema({})                                            |

The `llm.Tool` class has the following user-defined methods:

#### `async_call`

> Perform the actual operation of the tool when called by the LLM. This must be an async method and return json-serializable result. Its arguments are `hass` and an instance of `llm.ToolInput`.

Reviewer: I think it would be good to have similar to https://developers.home-assistant.io/docs/dev_101_services/#response-data (1) a more specific definition of the response type and (2)  More specific definition of how errors are handled.

Below you have "json-serializable result." which is good and so that offers a way to be more specific here.

| `user_prompt`     | string  | The raw text input that initiated the tool call                                                         |
| `language`        | string  | The language of the conversation agent, or "*" for any language                                         |
| `agent_id`        | string  | The ID of the conversation agent                                                                        |
| `conversation_id` | string  | The ID of the conversation                                                                              |
| `device_id`       | string  | The device ID that user uses for the conversation, if available                                         |
| `assistant`       | string  | The assistant name used to control exposed entities. Currently only `conversation` is supported.        |

### Unregistration
When your integration no longer wants to provide the tool, you need to unregister it with `llm.async_remove_tool`.

Example:
```python
async_remove_tool(hass, MyTool)
```

You can also use the tool name:
```python
async_remove_tool(hass, "multiply")
```

> ### Full example

Reviewer: There is already effectively a full example above for `MyTool` so I don't think both are needed and one should be removed or combined.

Author: You are right, this example used to be more extensive before I removed `custom_serializer`.

from homeassistant.helpers import llm

def currency_validator(currency_code: str) -> Currency:
    """Validate currency code."""
    return Currency(currency_code)

class MyTool(llm.Tool):
    """LLM Tool to get currency exponent."""

    name = "currency_exponent"
    description = "Return the currency exponent for a ISO 4217 currency code"
    parameters = vol.Schema(
        {
            vol.Required(
                "currency", description="ISO 4217 alphabetic code"
            ): currency_validator,
        }
    )

    async def async_call(self, hass: HomeAssistant, tool_input: llm.ToolInput) -> Any:
        """Call the function."""
        return {"exponent": tool_input.tool_args["currency"].exponent}

llm.async_register_tool(hass, MyTool())
```

## Option 3. Tool function

Define a function and register it as an LLM tool. This is the easiest method.

There are two ways to register the function as an LLM tool: with `llm.FunctionTool` class or with a decorator.

Example with `llm.FunctionTool` class:

```python
from homeassistant.helpers import llm

def multiply(a: int, b: int) -> dict[str, int]:
    "Return a product of two integers."
    return {"result": a * b}

> my_tool = llm.FunctionTool(multiply)

Reviewer: Can you make these more realistic examples like in https://developers.home-assistant.io/docs/dev_101_services/ or  https://developers.home-assistant.io/docs/intent_handling

llm.async_register_tool(hass, my_tool)
```

Example with the `@llm_tool` decorator:
```python
from homeassistant.helpers.llm import llm_tool

> @llm_tool(hass)

Reviewer: This example doesn't really show where `hass` comes from and given its only useful for inline functions where `hass` came from some other function, i'm not sure its adding value as a decorator over the previous example.
